// toppy/esm5/lib/position/global-position.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { InsidePlacement } from '../models';
import { setWH } from '../utils';
import { ToppyPosition } from './position';
/**
 * @record
 */
function GlobalPositionConfig() { }
if (false) {
    /** @type {?|undefined} */
    GlobalPositionConfig.prototype.placement;
    /** @type {?|undefined} */
    GlobalPositionConfig.prototype.offset;
    /** @type {?|undefined} */
    GlobalPositionConfig.prototype.width;
    /** @type {?|undefined} */
    GlobalPositionConfig.prototype.height;
}
var GlobalPosition = /** @class */ (function (_super) {
    tslib_1.__extends(GlobalPosition, _super);
    function GlobalPosition(config) {
        var _this = _super.call(this) || this;
        _this.config = { placement: InsidePlacement.CENTER, width: 100, height: 100, offset: 0 };
        _this.updateConfig(config);
        return _this;
    }
    /**
     * @param {?=} hostEl
     * @return {?}
     */
    GlobalPosition.prototype.getPositions = /**
     * @param {?=} hostEl
     * @return {?}
     */
    function (hostEl) {
        /** @type {?} */
        var host = (/** @type {?} */ (hostEl.getBoundingClientRect()));
        /** @type {?} */
        var src = {
            width: window['innerWidth'],
            height: window['innerHeight']
        };
        var _a = this.config, w = _a.width, h = _a.height;
        w = setWH(src, host, 'width', w);
        h = setWH(src, host, 'height', h);
        /** @type {?} */
        var props = this.calc(this.config.placement, src, host);
        return tslib_1.__assign({}, props, { width: w, height: h, position: 'fixed', extra: this.config.placement });
    };
    /**
     * @private
     * @param {?} placement
     * @param {?} src
     * @param {?} host
     * @return {?}
     */
    GlobalPosition.prototype.calc = /**
     * @private
     * @param {?} placement
     * @param {?} src
     * @param {?} host
     * @return {?}
     */
    function (placement, src, host) {
        var _a = tslib_1.__read(placement.split(''), 2), main = _a[0], sub = _a[1];
        /** @type {?} */
        var p = {};
        if (main === 't') {
            p.top = this.config.offset;
        }
        if (main === 'b') {
            p.bottom = this.config.offset;
        }
        if ((main === 'l' || main === 'r' || main === 'c') && !sub) {
            p.top = (src.height - host.height) / 2;
        }
        if ((main === 't' || main === 'b' || main === 'c') && !sub) {
            p.left = (src.width - host.width) / 2;
        }
        if ((main === 'l' && !sub) || sub === 'l') {
            p.left = this.config.offset;
        }
        if ((main === 'r' && !sub) || sub === 'r') {
            p.right = this.config.offset;
        }
        return p;
    };
    return GlobalPosition;
}(ToppyPosition));
export { GlobalPosition };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    GlobalPosition.prototype.config;
}
