// toppy/esm5/lib/position/position.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/**
 * @abstract
 */
var /**
 * @abstract
 */
ToppyPosition = /** @class */ (function () {
    function ToppyPosition() {
        this.config = {};
    }
    /**
     * @return {?}
     */
    ToppyPosition.prototype.getClassName = /**
     * @return {?}
     */
    function () {
        return this.constructor.name.replace('Pos', '-pos').toLowerCase();
    };
    /**
     * @param {?} config
     * @return {?}
     */
    ToppyPosition.prototype.updateConfig = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        this.config = tslib_1.__assign({}, this.config, config);
    };
    /**
     * @param {?} tid
     * @return {?}
     */
    ToppyPosition.prototype.init = /**
     * @param {?} tid
     * @return {?}
     */
    function (tid) { };
    return ToppyPosition;
}());
/**
 * @abstract
 */
export { ToppyPosition };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    ToppyPosition.prototype.config;
    /**
     * @abstract
     * @param {?} host
     * @return {?}
     */
    ToppyPosition.prototype.getPositions = function (host) { };
}
