// toppy/esm5/lib/toppy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ApplicationRef, ComponentFactoryResolver, Injectable, Injector } from '@angular/core';
import { DefaultConfig } from './config';
import { InsidePlacement } from './models';
import { GlobalPosition } from './position';
import { ToppyControl } from './toppy-control';
import { Bus, createId, getContent } from './utils';
import * as i0 from "@angular/core";
var Toppy = /** @class */ (function () {
    function Toppy(injector) {
        this.injector = injector;
        this.inputs = {
            position: null,
            config: DefaultConfig,
            content: { type: "s" /* STRING */, data: 'hello', props: {} },
            tid: null
        };
        this.inputs.position = new GlobalPosition({ placement: InsidePlacement.TOP });
    }
    /**
     * @param {?} position
     * @return {?}
     */
    Toppy.prototype.position = /**
     * @param {?} position
     * @return {?}
     */
    function (position) {
        this.inputs.position = position;
        return this;
    };
    /**
     * @param {?} config
     * @return {?}
     */
    Toppy.prototype.config = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        this.inputs.config = tslib_1.__assign({}, DefaultConfig, config);
        return this;
    };
    /**
     * @param {?} data
     * @param {?=} props
     * @return {?}
     */
    Toppy.prototype.content = /**
     * @param {?} data
     * @param {?=} props
     * @return {?}
     */
    function (data, props) {
        if (props === void 0) { props = {}; }
        this.inputs.content = getContent(data, props);
        return this;
    };
    /**
     * @param {?=} key
     * @return {?}
     */
    Toppy.prototype.create = /**
     * @param {?=} key
     * @return {?}
     */
    function (key) {
        if (key === void 0) { key = null; }
        this.tid = this.inputs.tid = key || createId();
        /** @type {?} */
        var injector = Injector.create([
            {
                provide: ToppyControl,
                deps: [ApplicationRef, ComponentFactoryResolver, Injector]
            }
        ], this.injector);
        /** @type {?} */
        var tc = injector.get(ToppyControl);
        if (Toppy.controls[this.tid]) {
            this.tid = createId();
        }
        this.inputs.position.init(this.tid);
        Toppy.controls[this.tid] = Object.assign(tc, this.inputs);
        return tc;
    };
    /**
     * @param {?} tid
     * @return {?}
     */
    Toppy.prototype.getCtrl = /**
     * @param {?} tid
     * @return {?}
     */
    function (tid) {
        return Toppy.controls[tid];
    };
    /**
     * @return {?}
     */
    Toppy.prototype.destroy = /**
     * @return {?}
     */
    function () {
        // tslint:disable-next-line:forin
        for (var key in Toppy.controls) {
            Toppy.controls[key].close();
        }
        Toppy.controls = {};
        Bus.stop();
    };
    Toppy.controls = {};
    Toppy.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    Toppy.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    /** @nocollapse */ Toppy.ngInjectableDef = i0.defineInjectable({ factory: function Toppy_Factory() { return new Toppy(i0.inject(i0.INJECTOR)); }, token: Toppy, providedIn: "root" });
    return Toppy;
}());
export { Toppy };
if (false) {
    /** @type {?} */
    Toppy.controls;
    /**
     * @type {?}
     * @private
     */
    Toppy.prototype.tid;
    /**
     * @type {?}
     * @private
     */
    Toppy.prototype.inputs;
    /**
     * @type {?}
     * @private
     */
    Toppy.prototype.injector;
}
