// toppy/esm5/lib/utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { TemplateRef } from '@angular/core';
import { Subject } from 'rxjs';
import { filter, map } from 'rxjs/operators';
/**
 * @param {?} data
 * @param {?=} props
 * @return {?}
 */
export function getContent(data, props) {
    if (props === void 0) { props = {}; }
    /** @type {?} */
    var type = "c" /* COMPONENT */;
    if (typeof data === 'string' && props['hasHTML'])
        type = "h" /* HTML */;
    else if (typeof data === 'string')
        type = "s" /* STRING */;
    else if (data instanceof TemplateRef)
        type = "t" /* TEMPLATE */;
    return { data: data, type: type, props: props };
}
/**
 * @return {?}
 */
export function createId() {
    return Math.random()
        .toString(36)
        .substr(2, 5);
}
/* html dom utils */
/**
 * @param {?} method
 * @param {?} cls
 * @param {?=} target
 * @return {?}
 */
export function cssClass(method, cls, target) {
    if (target === void 0) { target = 'body'; }
    var _a;
    (_a = document.querySelector(target).classList)[method].apply(_a, tslib_1.__spread(cls));
}
/**
 * @param {?} styleObj
 * @return {?}
 */
export function toCss(styleObj) {
    return Object.keys(styleObj)
        .map(function (x) { return x + ":" + styleObj[x] + (typeof styleObj[x] === 'number' ? 'px' : ''); })
        .join(';');
}
/**
 * @param {?} max
 * @param {?} percentage
 * @return {?}
 */
export function percentToCss(max, percentage) {
    /** @type {?} */
    var number = Number(percentage.slice(0, -1));
    if (number > 100) {
        number = 100;
    }
    return "calc(" + max + "px - " + (100 - number) + "%)";
}
/**
 * @param {?} src
 * @param {?} host
 * @param {?} key
 * @param {?} value
 * @return {?}
 */
export function setWH(src, host, key, value) {
    if (typeof value === 'number') {
        host[key] = value = Math.abs(value);
    }
    if (typeof value === 'string' && value.endsWith('%')) {
        value = percentToCss(src[key], value);
    }
    return value;
}
/** @type {?} */
export var BodyEl = document.querySelector('body');
/* events */
var /* events */
BusClass = /** @class */ (function () {
    function BusClass() {
        this._e = new Subject();
    }
    /**
     * @param {?} from
     * @param {?} name
     * @param {?=} data
     * @return {?}
     */
    BusClass.prototype.send = /**
     * @param {?} from
     * @param {?} name
     * @param {?=} data
     * @return {?}
     */
    function (from, name, data) {
        if (data === void 0) { data = null; }
        this._e.next({ from: from, name: name, data: data });
    };
    /**
     * @param {?} from
     * @param {?} name
     * @return {?}
     */
    BusClass.prototype.listen = /**
     * @param {?} from
     * @param {?} name
     * @return {?}
     */
    function (from, name) {
        return this._e.asObservable().pipe(filter(function (e) { return e.from === from && e.name === name; }), map(function (e) { return e.data; }));
    };
    /**
     * @return {?}
     */
    BusClass.prototype.stop = /**
     * @return {?}
     */
    function () {
        this._e.complete();
    };
    return BusClass;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    BusClass.prototype._e;
}
/** @type {?} */
export var Bus = new BusClass();
