// toppy/esm2015/lib/models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PositionMeta() { }
if (false) {
    /** @type {?|undefined} */
    PositionMeta.prototype.top;
    /** @type {?|undefined} */
    PositionMeta.prototype.left;
    /** @type {?|undefined} */
    PositionMeta.prototype.bottom;
    /** @type {?|undefined} */
    PositionMeta.prototype.right;
    /** @type {?|undefined} */
    PositionMeta.prototype.height;
    /** @type {?|undefined} */
    PositionMeta.prototype.width;
    /** @type {?|undefined} */
    PositionMeta.prototype.position;
    /** @type {?|undefined} */
    PositionMeta.prototype.extra;
}
/** @enum {string} */
const p = {
    TOP: 't',
    LEFT: 'l',
    RIGHT: 'r',
    BOTTOM: 'b',
    TOP_LEFT: 'tl',
    TOP_RIGHT: 'tr',
    BOTTOM_LEFT: 'bl',
    BOTTOM_RIGHT: 'br',
};
/** @enum {string} */
const o = {
    LEFT_TOP: 'lt',
    RIGHT_TOP: 'rt',
    LEFT_BOTTOM: 'lb',
    RIGHT_BOTTOM: 'rb',
};
/** @enum {string} */
const i = {
    CENTER: 'c',
};
/** @type {?} */
export const OutsidePlacement = Object.assign({}, p, o);
/** @type {?} */
export const InsidePlacement = Object.assign({}, p, i);
/** @enum {string} */
const SlidePlacement = {
    LEFT: 'l',
    RIGHT: 'r',
};
export { SlidePlacement };
/**
 * @record
 */
export function ContainerSize() { }
if (false) {
    /** @type {?} */
    ContainerSize.prototype.width;
    /** @type {?} */
    ContainerSize.prototype.height;
}
/**
 * @record
 */
export function ToppyConfig() { }
if (false) {
    /** @type {?} */
    ToppyConfig.prototype.backdrop;
    /** @type {?} */
    ToppyConfig.prototype.containerClass;
    /** @type {?} */
    ToppyConfig.prototype.wrapperClass;
    /** @type {?} */
    ToppyConfig.prototype.backdropClass;
    /** @type {?} */
    ToppyConfig.prototype.listenWindowEvents;
    /** @type {?} */
    ToppyConfig.prototype.closeOnDocClick;
    /** @type {?} */
    ToppyConfig.prototype.bodyClass;
    /** @type {?} */
    ToppyConfig.prototype.closeOnEsc;
    /** @type {?} */
    ToppyConfig.prototype.windowResizeCallback;
    /** @type {?} */
    ToppyConfig.prototype.docClickCallback;
}
/**
 * @record
 * @template T
 */
export function ComponentType() { }
/**
 * @record
 */
export function ToppyEvent() { }
if (false) {
    /** @type {?} */
    ToppyEvent.prototype.from;
    /** @type {?} */
    ToppyEvent.prototype.name;
    /** @type {?|undefined} */
    ToppyEvent.prototype.data;
}
/** @enum {string} */
const ContentType = {
    STRING: 's',
    HTML: 'h',
    TEMPLATE: 't',
    COMPONENT: 'c',
};
export { ContentType };
/**
 * @record
 */
export function Content() { }
if (false) {
    /** @type {?|undefined} */
    Content.prototype.type;
    /** @type {?} */
    Content.prototype.data;
    /** @type {?|undefined} */
    Content.prototype.props;
}
/**
 * @record
 */
export function Inputs() { }
if (false) {
    /** @type {?} */
    Inputs.prototype.position;
    /** @type {?} */
    Inputs.prototype.config;
    /** @type {?} */
    Inputs.prototype.content;
    /** @type {?} */
    Inputs.prototype.tid;
}
