// toppy/esm2015/lib/toppy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ApplicationRef, ComponentFactoryResolver, Injectable, Injector } from '@angular/core';
import { DefaultConfig } from './config';
import { InsidePlacement } from './models';
import { GlobalPosition } from './position';
import { ToppyControl } from './toppy-control';
import { Bus, createId, getContent } from './utils';
import * as i0 from "@angular/core";
export class Toppy {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
        this.inputs = {
            position: null,
            config: DefaultConfig,
            content: { type: "s" /* STRING */, data: 'hello', props: {} },
            tid: null
        };
        this.inputs.position = new GlobalPosition({ placement: InsidePlacement.TOP });
    }
    /**
     * @param {?} position
     * @return {?}
     */
    position(position) {
        this.inputs.position = position;
        return this;
    }
    /**
     * @param {?} config
     * @return {?}
     */
    config(config) {
        this.inputs.config = Object.assign({}, DefaultConfig, config);
        return this;
    }
    /**
     * @param {?} data
     * @param {?=} props
     * @return {?}
     */
    content(data, props = {}) {
        this.inputs.content = getContent(data, props);
        return this;
    }
    /**
     * @param {?=} key
     * @return {?}
     */
    create(key = null) {
        this.tid = this.inputs.tid = key || createId();
        /** @type {?} */
        const injector = Injector.create([
            {
                provide: ToppyControl,
                deps: [ApplicationRef, ComponentFactoryResolver, Injector]
            }
        ], this.injector);
        /** @type {?} */
        const tc = injector.get(ToppyControl);
        if (Toppy.controls[this.tid]) {
            this.tid = createId();
        }
        this.inputs.position.init(this.tid);
        Toppy.controls[this.tid] = Object.assign(tc, this.inputs);
        return tc;
    }
    /**
     * @param {?} tid
     * @return {?}
     */
    getCtrl(tid) {
        return Toppy.controls[tid];
    }
    /**
     * @return {?}
     */
    destroy() {
        // tslint:disable-next-line:forin
        for (const key in Toppy.controls) {
            Toppy.controls[key].close();
        }
        Toppy.controls = {};
        Bus.stop();
    }
}
Toppy.controls = {};
Toppy.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
Toppy.ctorParameters = () => [
    { type: Injector }
];
/** @nocollapse */ Toppy.ngInjectableDef = i0.defineInjectable({ factory: function Toppy_Factory() { return new Toppy(i0.inject(i0.INJECTOR)); }, token: Toppy, providedIn: "root" });
if (false) {
    /** @type {?} */
    Toppy.controls;
    /**
     * @type {?}
     * @private
     */
    Toppy.prototype.tid;
    /**
     * @type {?}
     * @private
     */
    Toppy.prototype.inputs;
    /**
     * @type {?}
     * @private
     */
    Toppy.prototype.injector;
}
