// toppy/esm2015/lib/utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { TemplateRef } from '@angular/core';
import { Subject } from 'rxjs';
import { filter, map } from 'rxjs/operators';
/**
 * @param {?} data
 * @param {?=} props
 * @return {?}
 */
export function getContent(data, props = {}) {
    /** @type {?} */
    let type = "c" /* COMPONENT */;
    if (typeof data === 'string' && props['hasHTML'])
        type = "h" /* HTML */;
    else if (typeof data === 'string')
        type = "s" /* STRING */;
    else if (data instanceof TemplateRef)
        type = "t" /* TEMPLATE */;
    return { data, type, props };
}
/**
 * @return {?}
 */
export function createId() {
    return Math.random()
        .toString(36)
        .substr(2, 5);
}
/* html dom utils */
/**
 * @param {?} method
 * @param {?} cls
 * @param {?=} target
 * @return {?}
 */
export function cssClass(method, cls, target = 'body') {
    document.querySelector(target).classList[method](...cls);
}
/**
 * @param {?} styleObj
 * @return {?}
 */
export function toCss(styleObj) {
    return Object.keys(styleObj)
        .map(x => `${x}:${styleObj[x]}${typeof styleObj[x] === 'number' ? 'px' : ''}`)
        .join(';');
}
/**
 * @param {?} max
 * @param {?} percentage
 * @return {?}
 */
export function percentToCss(max, percentage) {
    /** @type {?} */
    let number = Number(percentage.slice(0, -1));
    if (number > 100) {
        number = 100;
    }
    return `calc(${max}px - ${100 - number}%)`;
}
/**
 * @param {?} src
 * @param {?} host
 * @param {?} key
 * @param {?} value
 * @return {?}
 */
export function setWH(src, host, key, value) {
    if (typeof value === 'number') {
        host[key] = value = Math.abs(value);
    }
    if (typeof value === 'string' && value.endsWith('%')) {
        value = percentToCss(src[key], value);
    }
    return value;
}
/** @type {?} */
export const BodyEl = document.querySelector('body');
/* events */
class BusClass {
    constructor() {
        this._e = new Subject();
    }
    /**
     * @param {?} from
     * @param {?} name
     * @param {?=} data
     * @return {?}
     */
    send(from, name, data = null) {
        this._e.next({ from, name, data });
    }
    /**
     * @param {?} from
     * @param {?} name
     * @return {?}
     */
    listen(from, name) {
        return this._e.asObservable().pipe(filter(e => e.from === from && e.name === name), map(e => e.data));
    }
    /**
     * @return {?}
     */
    stop() {
        this._e.complete();
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    BusClass.prototype._e;
}
/** @type {?} */
export const Bus = new BusClass();
