// toppy/esm2015/lib/position/global-position.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InsidePlacement } from '../models';
import { setWH } from '../utils';
import { ToppyPosition } from './position';
/**
 * @record
 */
function GlobalPositionConfig() { }
if (false) {
    /** @type {?|undefined} */
    GlobalPositionConfig.prototype.placement;
    /** @type {?|undefined} */
    GlobalPositionConfig.prototype.offset;
    /** @type {?|undefined} */
    GlobalPositionConfig.prototype.width;
    /** @type {?|undefined} */
    GlobalPositionConfig.prototype.height;
}
export class GlobalPosition extends ToppyPosition {
    /**
     * @param {?} config
     */
    constructor(config) {
        super();
        this.config = { placement: InsidePlacement.CENTER, width: 100, height: 100, offset: 0 };
        this.updateConfig(config);
    }
    /**
     * @param {?=} hostEl
     * @return {?}
     */
    getPositions(hostEl) {
        /** @type {?} */
        const host = (/** @type {?} */ (hostEl.getBoundingClientRect()));
        /** @type {?} */
        const src = {
            width: window['innerWidth'],
            height: window['innerHeight']
        };
        let { width: w, height: h } = this.config;
        w = setWH(src, host, 'width', w);
        h = setWH(src, host, 'height', h);
        /** @type {?} */
        const props = this.calc(this.config.placement, src, host);
        return Object.assign({}, props, { width: w, height: h, position: 'fixed', extra: this.config.placement });
    }
    /**
     * @private
     * @param {?} placement
     * @param {?} src
     * @param {?} host
     * @return {?}
     */
    calc(placement, src, host) {
        const [main, sub] = placement.split('');
        /** @type {?} */
        const p = {};
        if (main === 't') {
            p.top = this.config.offset;
        }
        if (main === 'b') {
            p.bottom = this.config.offset;
        }
        if ((main === 'l' || main === 'r' || main === 'c') && !sub) {
            p.top = (src.height - host.height) / 2;
        }
        if ((main === 't' || main === 'b' || main === 'c') && !sub) {
            p.left = (src.width - host.width) / 2;
        }
        if ((main === 'l' && !sub) || sub === 'l') {
            p.left = this.config.offset;
        }
        if ((main === 'r' && !sub) || sub === 'r') {
            p.right = this.config.offset;
        }
        return p;
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    GlobalPosition.prototype.config;
}
